### ISSUE TITLE
Cryptsetup: please compile without --disable-kernel_crypto

### ISSUE THREAD

**Issue by blackfuel** [Entware-ng/Entware-ng#585]
### Package: Cryptsetup
I created an encrypted file container using Veracrypt 1.19, but it cannot be mounted on the router with Cryptsetup.  When I run the following test script, Cryptsetup refuses to open my Veracrypt encrypted device.  The error is, `Required kernel crypto interface not available`.  

To fix it, you'll need to compile without `--disable-kernel_crypto`.  AND, you'll need one header file from the Linux 2.6.38.8 source code distribution.  Just need to copy `linux/if_alg.h` from Linux 2.6.38.8, to `~/Entware-ng/staging_dir/target-arm_cortex-a9_glibc-2.23_eabi/usr/include/linux/if_alg.h`.  That's it, it's fixed.

Get the kernel header file, `linux/if_alg.h`, here:
https://cdn.kernel.org/pub/linux/kernel/v2.6/linux-2.6.38.8.tar.gz
  or
[if_alg.h.txt](https://github.com/Entware-ng/Entware-ng/files/820148/if_alg.h.txt)

To verify that this fix actually works, I had to backport the user-space crypto interface from Linux 2.6.38.8 to Asuswrt-Merlin.  My patch fixes all three ARM kernels in Asuswrt-Merlin.  *You only need to apply the kernel patch if you intend to test it yourself.*

My test script is this:
```
# first, use Veracrypt to create an encrypted file container named "vctest.img"
# then copy it to the router and run this script
losetup /dev/loop6 ./vctest.img
modprobe algif_skcipher  # you'll need my kernel patch (see below)
cryptsetup --veracrypt tcryptOpen /dev/loop6 vctest
# now wait 5 minutes; yes, it takes a very long time
mkdir /mnt/vctest
mount /dev/mapper/vctest /mnt/vctest
# now copy some files that you want encrypted, to the /mnt/vctest folder
umount /mnt/vctest
cryptsetup tcryptClose vctest
losetup -d /dev/loop6
```

My router is RT-AC68U (ARM) with Asuswrt-Merlin 380.66-alpha1.  I'm using Windows 10 and Veracrypt 1.19 to create a 32MB encrypted file container with FAT as the underlying volume format.  I also verified this works with Truecrypt hidden volumes with ext3 as the underlying volume format.  It works perfectly on the RT-AC68U router with Asuswrt-Merlin.

Patch to add the `algif_skcipher` kernel module to Asuswrt-Merlin firmware:
[380.66-alpha1_ARM_dm-crypt+skcipher.patch.txt](https://github.com/Entware-ng/Entware-ng/files/820222/380.66-alpha1_ARM_dm-crypt.skcipher.patch.txt)

How to apply the patch:
```
cd ~/asuswrt-merlin
patch -p2 -i ~/380.66-alpha1_ARM_dm-crypt+skcipher.patch
```

**EDIT: The attached patch file is now fixed.  It wasn't copying the new kernel modules to the firmware image.**

**Comment by zyxmon** [comment]
Entware is a repo for many different devices. We are not going to apply specific kernel backports or hacks for specific devices. Sorry.

**Comment by blackfuel** [comment]
My request is to support all Linux versions and all platforms for this feature of Cryptsetup.  There is no kernel backport or hack to worry about.   I was explaining how I tested it and verified that it works.  If your test platform is higher than Linux 2.6.38, then your testing work it's easy.  I will also create a request for mainline OpenWRT.  Hopefully they respond.

Compiling in `if_alg.h` allows Cryptsetup to communicate with the Linux kernel via user-space socket interface.  The `if_alg.h` header file was introduced in Linux 2.6.38 and hasn't changed in Linux 4.10.

In the Linux 4.10 kernel, the file has moved but didn't change.
https://cdn.kernel.org/pub/linux/kernel/v4.x/linux-4.4.10.tar.gz
Moved to a new place:
`./linux-4.10.1/include/uapi/linux/if_alg.h`

If you put this file in the following location, then the Cryptsetup makefile can see it when the `--disable-kernel_crypto` switch is removed.
`~/Entware-ng/staging_dir/target-arm_cortex-a9_glibc-2.23_eabi/usr/include/linux/if_alg.h`
